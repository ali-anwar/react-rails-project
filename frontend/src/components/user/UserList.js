import React from 'react';
import PropTypes from 'prop-types';
import { BootstrapTable, TableHeaderColumn } from 'react-bootstrap-table';
import _ from 'lodash';

const getCaret = direction => {
  if (direction === 'asc') {
    return (
      <span>
        {' '}
        <i className="fa fa-sort-asc" aria-hidden="true" />
      </span>
    );
  }

  if (direction === 'desc') {
    return (
      <span>
        {' '}
        <i className="fa fa-sort-desc" aria-hidden="true" />
      </span>
    );
  }

  return (
    <span>
      {' '}
      <i className="fa fa-sort" aria-hidden="true" />
    </span>
  );
};

const roleFormatter = (cell, row) => {
  return `${_.startCase(_.toLower(cell))}`;
};

const buttonFormatter = (cell, row, props) => {
  return (
    <div>
      <button
        className="btn btn-warning mr-2"
        onClick={() => props.handleEditButton(row.id)}
      >
        <i className="fa fa-pencil" aria-hidden="true" />
      </button>

      <button
        className="btn btn-danger"
        onClick={() => props.handleDeleteButton(row.id)}
      >
        <i className="fa fa-trash" aria-hidden="true" />
      </button>
    </div>
  );
};

class UserList extends React.Component {
  constructor(props) {
    super(props);

    this.options = {
      sortIndicator: true,
      noDataText: 'No data'
    };

    this.selectRowProp = {
      mode: 'radio',
      onSelect: props.handleRowSelect,
      clickToSelect: true,
      hideSelectColumn: true
    };
  }

  render() {
    return (
      <BootstrapTable
        data={this.props.users}
        selectRow={this.selectRowProp}
        options={this.options}
        bordered={false}
        className="custom-table"
        striped
        hover
        condensed
      >
        <TableHeaderColumn dataField="id" isKey hidden>
          Id
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="name"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          Name
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="email"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          Email
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="role"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
          dataFormat={roleFormatter}
        >
          Role
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="button"
          dataFormat={buttonFormatter}
          formatExtraData={this.props}
        >
          Actions
        </TableHeaderColumn>
      </BootstrapTable>
    );
  }
}

UserList.propTypes = {
  users: PropTypes.array.isRequired,
  handleRowSelect: PropTypes.func.isRequired,
  handleDeleteButton: PropTypes.func.isRequired,
  handleEditButton: PropTypes.func.isRequired
};

export default UserList;
