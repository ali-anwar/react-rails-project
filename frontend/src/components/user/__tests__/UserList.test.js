import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import UserList from '../UserList';


describe('UserList.test.js', () => {

  it('renders without crashing', () => {
    const props = {
      users: [{
        id: "react-flux-building-applications",
        title: "Building Applications in React and Flux",
        calories: 100,
        description: "Very Tasty"
      }],
      handleRowSelect: jest.fn(),
      handleDeleteButton: jest.fn(),
      handleEditButton: jest.fn(),
    };
    
    const wrapper = shallow(<UserList {...props}/>);

    expect(wrapper).toHaveLength(1);
    expect(wrapper.find('BootstrapTable')).toHaveLength(1);

    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
