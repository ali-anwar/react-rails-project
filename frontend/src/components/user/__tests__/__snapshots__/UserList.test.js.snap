exports[`UserList.test.js renders without crashing 1`] = `
<BootstrapTable
  bodyContainerClass={null}
  bordered={false}
  cellEdit={
    Object {
      "afterSaveCell": undefined,
      "beforeSaveCell": undefined,
      "blurToSave": false,
      "mode": "none",
      "nonEditableRows": undefined,
    }
  }
  className="custom-table"
  columnFilter={false}
  condensed={true}
  containerClass={null}
  csvFileName="spreadsheet.csv"
  data={
    Array [
      Object {
        "calories": 100,
        "description": "Very Tasty",
        "id": "react-flux-building-applications",
        "title": "Building Applications in React and Flux",
      },
    ]
  }
  deleteRow={false}
  expandColumnOptions={
    Object {
      "expandColumnBeforeSelectColumn": true,
      "expandColumnComponent": undefined,
      "expandColumnVisible": false,
    }
  }
  exportCSV={false}
  fetchInfo={
    Object {
      "dataTotalSize": 0,
    }
  }
  headerContainerClass={null}
  height="100%"
  hover={true}
  ignoreSinglePage={false}
  insertRow={false}
  keyBoardNav={false}
  multiColumnSearch={false}
  multiColumnSort={1}
  options={
    Object {
      "noDataText": "No data",
      "sortIndicator": true,
    }
  }
  pagination={false}
  printable={false}
  replace={false}
  search={false}
  selectRow={
    Object {
      "clickToSelect": true,
      "hideSelectColumn": true,
      "mode": "radio",
      "onSelect": [Function],
    }
  }
  striped={true}
  tableBodyClass={null}
  tableContainerClass={null}
  tableHeaderClass={null}
  trClassName=""
  withoutTabIndex={false}>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    dataAlign="left"
    dataField="id"
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={true}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={true}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Id
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="name"
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Name
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="email"
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Email
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="role"
    dataFormat={[Function]}
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Role
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    dataAlign="left"
    dataField="button"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    formatExtraData={
      Object {
        "handleDeleteButton": [Function],
        "handleEditButton": [Function],
        "handleRowSelect": [Function],
        "users": Array [
          Object {
            "calories": 100,
            "description": "Very Tasty",
            "id": "react-flux-building-applications",
            "title": "Building Applications in React and Flux",
          },
        ],
      }
    }
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Actions
  </TableHeaderColumn>
</BootstrapTable>
`;
